swiftide/src/integrations/treesitter/summarizer.rs

```diff
@@ -0,0 +1,297 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `code` - The source code to be split.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<String>` - A result containing a string, or an error if the code could not be parsed.
+    ///
+    /// # Errors
+    /// * If the code could not be parsed, an error is returned.
+    pub fn summarize(&self, code: &str) -> Result<String> {
+        let mut parser = Parser::new();
+        parser.set_language(&self.language.into())?;
+        let tree = parser.parse(code, None).context("No nodes found")?;
+        let root_node = tree.root_node();
+
+        if root_node.has_error() {
+            anyhow::bail!("Root node has invalid syntax");
+        }
+
+        let mut cursor = root_node.walk();
+        let mut summary = String::with_capacity(code.len());
+        let mut last_end = 0;
+        self.summarize_node(&mut cursor, code, &mut summary, &mut last_end);
+        Ok(summary)
+    }
+
+    fn is_unneeded_node(&self, node: Node) -> bool {
+        // We can use self.language to determine if a node is needed
+        match self.language {
+            SupportedLanguages::Rust => matches!(node.kind(), "block"),
+            SupportedLanguages::Typescript | SupportedLanguages::Javascript => {
+                matches!(node.kind(), "statement_block")
+            }
+            SupportedLanguages::Python => match node.kind() {
+                "block" => {
+                    let parent = node.parent().expect("Python block node has no parent");
+                    parent.kind() == "function_definition"
+                }
+                _ => false,
+            },
+            SupportedLanguages::Ruby => match node.kind() {
+                "body_statement" => {
+                    let parent = node
+                        .parent()
+                        .expect("Ruby body_statement node has no parent");
+                    parent.kind() == "def"
+                }
+                _ => false,
+            },
+        }
+    }
+
+    /// Summarizes a syntax node
+    ///
+    /// # Arguments
+    ///
+    /// * `node` - The syntax node to be chunked.
+    /// * `source` - The source code as a string.
+    /// * `last_end` - The end byte of the last chunk.
+    ///
+    /// # Returns
+    ///
+    /// * `String` - A summary of the syntax node.
+    fn summarize_node(
+        &self,
+        cursor: &mut TreeCursor,
+        source: &str,
+        summary: &mut String,
+        last_end: &mut usize,
+    ) {
+        let node = cursor.node();
+        // If the node is not needed in the summary, skip it and go to the next sibling
+        if self.is_unneeded_node(node) {
+            summary.push_str(&source[*last_end..node.start_byte()]);
+            *last_end = node.end_byte();
+            if cursor.goto_next_sibling() {
+                self.summarize_node(cursor, source, summary, last_end);
+            }
+            return;
+        }
+
+        let mut next_cursor = cursor.clone();
```

Comment:
Weird, does it really need a clone? The borrow should be returned after the block above since it's &mut

Reply:
Yes, it's actually a necessary clone not to satisfy the borrow checker, but because we need a clone, we don't want to modify the original value of the cursor because that cursor is needed to get us to the next sibling.